modules/genericAnalyticsAdapter.js

```diff
@@ -0,0 +1,155 @@
+  let batch = [];
+  let callDepth = 0;
+  let options, handler, timer, translate;
+
+  function optionsAreValid(options) {
+    if (!options.url && !options.handler) {
+      logError('options must specify either `url` or `handler`')
+      return false;
+    }
+    if (options.hasOwnProperty('method') && !['GET', 'POST'].includes(options.method)) {
+      logError('options.method must be GET or POST');
+      return false;
+    }
+    for (const [field, type] of Object.entries(TYPES)) {
+      // eslint-disable-next-line valid-typeof
+      if (options.hasOwnProperty(field) && typeof options[field] !== type) {
+        logError(`options.${field} must be a ${type}`);
+        return false;
+      }
+    }
+    if (options.hasOwnProperty('events')) {
+      if (!isPlainObject(options.events)) {
+        logError('options.events must be an object');
+        return false;
+      }
+      for (const [event, handler] of Object.entries(options.events)) {
+        if (!CONSTANTS.EVENTS.hasOwnProperty(event)) {
+          logWarn(`options.events.${event} does not match any known Prebid event`);
+          if (typeof handler !== 'function') {
+            logError(`options.events.${event} must be a function`);
+            return false;
+          }
+        }
+      }
+    }
+    return true;
+  }
+
+  function processBatch() {
+    const currentBatch = batch;
+    batch = [];
+    callDepth++;
+    try {
+      // the pub-provided handler may inadvertently cause an infinite chain of events;
+      // even just logging an exception from it may cause an AUCTION_DEBUG event, that
+      // gets back to the handler, that throws another exception etc.
+      // to avoid the issue, put a cap on recursion
+      if (callDepth === MAX_CALL_DEPTH) {
+        logError('detected probable infinite recursion, discarding events', currentBatch);
+      }
+      if (callDepth >= MAX_CALL_DEPTH) {
+        return;
+      }
+      try {
+        handler(currentBatch);
+      } catch (e) {
+        logError('error executing options.handler', e);
+      }
+    } finally {
+      callDepth--;
+    }
+  }
+
+  function translator(eventHandlers) {
+    if (!eventHandlers) {
+      return (data) => data;
+    }
+    return function ({eventType, args}) {
+      if (eventHandlers.hasOwnProperty(eventType)) {
+        try {
+          return eventHandlers[eventType](args);
+        } catch (e) {
+          logError(`error executing options.events.${eventType}`, e);
+        }
+      }
+    }
+  }
+
+  return Object.assign(
+    Object.create(parent),
+    {
+      enableAnalytics(config) {
+        if (optionsAreValid(config?.options || {})) {
+          options = Object.assign({}, DEFAULTS, config.options);
+          handler = options.handler || defaultHandler(options);
+          translate = translator(options.events);
+          parent.enableAnalytics.call(this, config);
+        }
+      },
+      track(event) {
```

Comment:
is it straightforward to see which events a particular analytics adapter subscribes to?